docs/appengine/firebase/tictactoe/main.go

```diff
@@ -0,0 +1,299 @@
+		"https://www.googleapis.com/auth/firebase.database",
+		"https://www.googleapis.com/auth/userinfo.email",
+	)
+	if err != nil {
+		return nil, err
+	}
+	base := os.Getenv("FIREBASE_BASE")
+	if base == "" {
+		// Check the environment variable for the base firebase URL.
+		//
+		// The config should look like:
+		//
+		// env_variables:
+		//    FIREBASE_BASE: https://app-id.firebase.io.com
+		//
+		return nil, errors.New("Missing FIREBASE_BASE environment variable.")
+	}
+	return firego.New(base, hc), nil
+}
+
+func moveHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	g, err := gameFromRequest(r)
+	if err != nil {
+		handleError(w, r, "Could not get game", err)
+		return
+	}
+
+	position, err := strconv.Atoi(r.FormValue("i"))
+	if err != nil {
+		handleError(w, r, "Invalid position", err)
+		return
+	}
+	if position < 0 || position > 8 {
+		handleError(w, r, "Expected position between 0 and 8", nil)
+		return
+	}
+
+	u := user.Current(ctx).ID
+	expectedUser := g.UserX
+	if !g.MoveX {
+		expectedUser = g.UserO
+	}
+	if u != expectedUser {
+		handleError(w, r, "Not your move", nil)
+		return
+	}
+
+	// Update the game board.
+	if err := g.MoveAt(position); err != nil {
+		handleError(w, r, "Could not move", err)
+	}
+	g.MoveX = !g.MoveX
+
+	if winner, isWon := g.CheckWin(); isWon {
+		if winner == "O" {
+			g.Winner = g.UserO
+		} else if winner == "X" {
+			g.Winner = g.UserX
+		} else {
+			g.Winner = "No one"
+		}
+		g.WinningBoard = g.Board // TODO: implement patterns
+	}
+
+	if _, err := datastore.Put(ctx, g.K, g); err != nil {
+		handleError(w, r, "Could not save game", err)
+		return
+	}
+	sendUpdate(ctx, g)
+}
+
+func deleteHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+
+	game, err := gameFromRequest(r)
+	if err != nil {
+		handleError(w, r, "Could not get game", err)
+		return
+	}
+	fb, err := firebase(ctx)
+
+	channelID := user.Current(ctx).ID + game.K.Encode()
+
+	if err := fb.Child("channels").Child(channelID).Remove(); err != nil {
+		handleError(w, r, "Could not delete channel", err)
+		return
+	}
+	log.Infof(ctx, "Deleted channel %v", channelID)
+}
+
+func openedHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+
+	k, err := datastore.DecodeKey(r.FormValue("g"))
+	if err != nil {
+		handleError(w, r, "Invalid game ID", err)
+		return
+	}
+	var g Game
+	if err := datastore.Get(ctx, k, &g); err != nil {
+		handleError(w, r, "Could not get game", err)
+		return
+	}
+	sendUpdate(ctx, &g)
+}
+
+func sendUpdate(ctx context.Context, g *Game) {
+	fb, err := firebase(ctx)
+	if err != nil {
+		log.Errorf(ctx, "getFirebase: %v", err)
+	}
+	chans := fb.Child("channels")
+
+	gameKey := g.K.Encode()
+
+	if g.UserO != "" {
+		channelID := g.UserO + gameKey
+		if err := chans.Child(channelID).Set(g); err != nil {
+			log.Errorf(ctx, "Updating UserO (%s): %v", channelID, err)
+		} else {
+			log.Infof(ctx, "Update O sent.")
```

Comment:
"Update O sent to %v", channelID?

Server will output many logs from all sorts of users. Identification is required if we think it is necessary to log the successful updates. If not, we shouldn't log.